### ISSUE TITLE
Feature request: support Linux ARM / ARM64

### ISSUE THREAD

**Issue by m-brooks** [dotnet/vscode-csharp#2771]
## Environment data
`dotnet --info` output:
.NET Core SDK (reflecting any global.json):
 Version:   2.2.101
 Commit:    236713b0b7

Runtime Environment:
 OS Name:     raspbian
 OS Version:  9
 OS Platform: Linux
 RID:         linux-arm
 Base Path:   /home/pi/dotnet/sdk/2.2.101/

Host (useful for support):
  Version: 2.2.0
  Commit:  1249f08fed

.NET Core SDKs installed:
  2.2.101 [/home/pi/dotnet/sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.2.0 [/home/pi/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.2.0 [/home/pi/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.2.0 [/home/pi/dotnet/shared/Microsoft.NETCore.App]

VS Code version:
Version: 1.29.0 (user setup)
Commit: f74c7914490b68b4052ff877f04b6eaf7198f261
Date: 2018-10-16T15:00:48.287Z
Electron: 2.0.11
Chrome: 61.0.3163.100
Node.js: 8.9.3
V8: 6.1.534.41
Architecture: arm

C# Extension version:
1.17.1

## Steps to reproduce
Hit F5 from a C# console application.

## Expected  behavior
Program builds and runs, and stops at breakpoints.

## Actual behavior
[WARNING]: Processor architecture 'armv7l' is not currently supported by the .NET Core debugger. Debugging will not be available.

**Comment by gregg-miskelly** [comment]
@m-brooks my understanding is that VS Code doesn't have an official version for Linux ARM. Is that correct?

**Comment by gregg-miskelly** [comment]
Short answer: This isn't something we can currently help with. The debugger is only licensed to run in official version of VS Code. So even if you had a version for ARM, it still wouldn't run because the license checks would fail. So https://github.com/Microsoft/vscode/issues/6442 needs to be addressed before this can be implemented.

I will leave this open in case the VS Code team decides to more officially support Linux ARM.

More details in case you are curious: The C# extension has 4 parts to it --

* The 'entry point' is a bit of JavaScript that interacts with the VS Code UI and glues everything together
* The language service is implemented in C# and runs on mono. It doesn't fully support ARM in the sense that it doesn't ship with an ARM version of mono. But from your description, it sounds like everything is fine as long as your box has a new enough ARM mono version installed.
* The debugger is implemented in a mixture of C# and C++. Most of this is available for Linux ARM because we support Linux ARM as a remote debugging target. But currently we don't package it together for VS Code since there isn't an ARM version of VS Code to support.
* The Razor language service is implemented in C#. There would need to be additional packaging work to support Linux ARM, but otherwise I believe things should just work.

**Comment by joaomoreno** [comment]
@gregg-miskelly Great news, the first bullet point has been addressed, since the latest Insiders already supports ARM as a remote SSH target: https://twitter.com/joaomoreno/status/1138379435839754240

Now all we need are the following 3 bullet points... 💪

**Comment by gregg-miskelly** [comment]
@joaomoreno very cool! But to be clear, I am not sure this changes things too much --

* If you want this because you just want to debug on a Raspberry Pi, but you will run VS Code on an x64 PC - that was actually something that the C# extension has supported for a long time via remote debugging. Some folks might find VS Code remote an interesting option as an alternative, but there are some downsides of that way too since you are running Omnisharp, which is pretty demanding, on ARM.
* If you want this because you have a very inexpensive ARM-based computer that you would like to use as your PC, then you probably need full VS Code support (and probably a .NET SDK)

**Comment by thesamesam** [comment]
@gregg-miskelly My use case is using a Pi as a base to connect to from other devices through the Remote feature, but even though vsdbg runs on the Pi, the extension doesn't seem to work (complains about architecture).

Given that a remote instance of Visual Studio Code would end up running the debugger, it would now be permitted under the license?

**Comment by gregg-miskelly** [comment]
@thesamesam there is no licensing reason why that scenario can't work, but I think you would be better off using [remote debugging](https://github.com/OmniSharp/omnisharp-vscode/wiki/Remote-Debugging-On-Linux-Arm) instead. Firstly, because it works today, so right now, it is the only option. But also because, at least if you have a reasonably sized project, the language service (Omnisharp) can be compute-hungry. So you probably would rather use your local processor, than run it on the comparatively under-powered CPU on the Pi.

**Comment by thesamesam** [comment]
@gregg-miskelly Oh, I misunderstood then. I can see why the processor might be a problem. The main reason I'm keen on it is having the filesystem constant when I'm changing machines.

At the moment, when I tried Remote (Insiders; Nightly SSH) + Pi + C# extension, I had the armv7 error for debugging and it didn't seem to download or fire up OmniSharp either (so nothing happened, bar the error). I'm not sure if I need to report this somewhere else though?

**Comment by gregg-miskelly** [comment]
If we fixed this issue, the remote scenario would also be addressed, so no need to open a separate issue.

**Comment by SuperJMN** [comment]
Big bummer this isn't working inside Raspbian. Cross-platform is still far from being real in 2019.

**Comment by NetGH** [comment]
By following Remote Development using SSH https://code.visualstudio.com/docs/remote/ssh

Python could be run and debugged without problem, 
However, run a c# project it displays

[ERROR] Error: spawn /home/pi/.vscode-server-insiders/extensions/ms-vscode.csharp-1.21.0/.omnisharp/1.34.0/run ENOENT

[Window Title]
Visual Studio Code - Insiders

[Content]
The C# extension is still downloading packages. Please see progress in the output window below.

[Open launch.json] [Cancel]

**Comment by gregg-miskelly** [comment]
@NetGH Correct. If you are trying to do remote ARM development, you want to use [remote debugging](https://github.com/OmniSharp/omnisharp-vscode/wiki/Remote-Debugging-On-Linux-Arm) instead. It would be better anyway...

**Comment by bpranger** [comment]
@gregg-miskelly I don't agree that "It would be better anyway".

I would like to be able to build and test C# programs using [System.Device.Gpio](https://github.com/dotnet/iot/tree/master/src/System.Device.Gpio) on a Raspberry Pi using VS Code and Remote-SSH.

It is especially onerous when I have to first create the projects on my Windows PC (in order to get language/IntelliSense support), then, at some point, copy my project to the rpi in order to test/debug (due to the lack of actual GPIO hardware on my Windows PC). Every time I need to make substantial edits, it's back to my Windows PC in order to leverage IntelliSense, then back over to rpi to test/debug; a dreadful cycle.

I am not concerned about taxing the rpi system resources a) because my performance requirements are not substantial, b) because I understand and can manage the tradeoffs, and c) because, frankly, newer rpi devices are relatively powerful.  

I see there is a "blocked on vs code" tag on this issue - is that still the case? I've read over this thread a few times and, from what I can see, there don't appear to be any show-stoppers; would this be difficult to enable? This seems like a slam-dunk feature request to me.

**Comment by nathansoz** [comment]
@gregg-miskelly I think there might be a use case now that supports enabling this. I have a Surface Pro X and want to use x86 code as a frontend into WSL (using the WSL extension). WSL is running native ARM64 so in theory running most of my development load there should be more efficient. Are there still blockers to doing this (given that omnisharp runs on mono and there exists an arm64 debugger now[1])?

[1]: https://github.com/OmniSharp/omnisharp-vscode/issues/2105

**Comment by gregg-miskelly** [comment]
@nathansoz interesting. So you are trying this now and WSL is supported on Surface Pro X? I would agree that this sounds like an interesting scenario. Unfortunately, I don't think you will be able to currently get it to work through the remote extension support - the remote debugger (which does support Linux arm64) is slightly smaller than what we use with VS Code. You can still do this via remote debugging if you like: https://github.com/OmniSharp/omnisharp-vscode/wiki/Windows-Subsystem-for-Linux

**Comment by nathansoz** [comment]
@gregg-miskelly Yeah, WSL v1 and v2 are both supported and work exactly as you would expect. Here's the output showing that it is running native ARM:

![image](https://user-images.githubusercontent.com/6821708/73395701-e3a42580-4294-11ea-9d2f-38fc9a50ce6d.png)

**Comment by tomflorin** [comment]
I'm using a Chromebook with crostini and on my aarch64 linux I can use the latest dotnet core that has arm64 support. Using f# and vscode is possible with the f# language server but it's sad to see that there is no good option for c#. I just need a decent intellisense and I can live without debugging, but I'm not sure what is missing and what can we do about it

**Comment by kangaroo** [comment]
This seems trivial to achieve:

https://gist.github.com/kangaroo/883b9a6e26c1c76780fccf41f23a629e

I just can't seem to figure out the canonical url, only the fallbackUrl.  That said, this change lets me debug C# projects on my rpi4 with the Remote SSH target without issue.

**Comment by sdmaclea** [comment]
MacOS Arm64 should increase the priority of this.

**Comment by Spongman** [comment]
so, basically we'd need to strip all the legacy cruft from Omnisharp in order to get it to run on dotnet core (again?!) and then this should work?

**Comment by lostmsu** [comment]
@kangaroo can you make a pull request?

**Comment by lostmsu** [comment]
You can now run a free (until the end of 2020) ARM64 instance on AWS: https://aws.amazon.com/blogs/aws/new-t4g-instances-burstable-performance-powered-by-aws-graviton2/

**Comment by jmevel** [comment]
Pinebook Pro user here. 

I understand the Omnisharp extension can only run on an official VS Code build.

Things are moving because very recently Microsoft announced the official Windows ARM version of VS Code.

The only thing I wish is that when an official Linux ARM version will be out, I'll be able to debug my favorite language on my `arm64` laptop.

Thanks to all the team for your work.

**Comment by baralong** [comment]
VS Code for arm is now released, currentky installed on my arm chromebook
```
Version: 1.50.0
Commit: 93c2f0fbf16c5a4b10e4d5f89737d9c2c25488a3
Date: 2020-10-07T06:01:12.076Z
Electron: 9.2.1
Chrome: 83.0.4103.122
Node.js: 12.14.1
V8: 8.3.110.13-electron.0
OS: Linux arm64 5.4.48-06434-gd868196d13af
```

**Comment by m-brooks** [comment]
It's now close to 2 years since I posted the OP on this thread and we've seen great strides. Microsoft now have official builds of VSCode for Linux ARM and I have just successfully installed the 32 bit version on my Raspberry Pi (the Pi itself uses a 64 bit CPU but the Raspian OS is still 32 bit). I have also installed both .Net Core 3.1 and the dotnet-script REPL. 

Unfortunately in one other respect things seem to have gone backwards. When I started this thread I had installed Jay Rodgers's  ARM port of VSCode and was successfully using the Microsoft C# extension on it. However, I have installed that extension on my Microsoft-built installation of VSCode only to find that it reports "The C# extension for Visual Studio Code (powered by OmniSharp) is incompatible on linux armv7l". 

That said, when I create a CSX file and start editing, all looks good (nice syntax highlighting) and formats nicely (I do also have the C# FixFormat 0.0.85 extension installed, so possibly that has something to do with it.

However, when I hit F5 to debug I get the message "Couldn't find a debug adapter descriptor for debug type 'coreclr' (extension might have failed to activate)", which seems to support the C# extension's assertion that it isn't working.

Anyone any insight into what needs to be done next in order to get C# or CSX debugging working?

**Comment by SLS8** [comment]
> if I type "for" then Intellisense suggests **for**, **foreach**, **forr** and I can make a selection and get a nice snippet, but if I type "Console." then Intellisense does not kick in to show me the methods and properties of the Console static class as it would on a Windows PC.

That's not Intellisense, that's the default autocomplete. It behaves like a simple text editor, it also suggests previously used words. I could live without a debugger, but without Intellisense what's the point of installing vscode

**Comment by scottyearsley** [comment]
Really hoping this gets some focus - I use VS Code and C# everywhere, but was sadly disappointed when I couldn't debug on my highly portable Lenovo Chromebook Duet

**Comment by baralong** [comment]
> Really hoping this gets some focus - I use VS Code and C# everywhere, but was sadly disappointed when I couldn't install on my highly portable Lenovo Chromebook Duet

I have the exact same machine. Hoping it comes soon.

**Comment by SuperJMN** [comment]
Raspbian user here wishing to have this feature.
________________________________

**Comment by baralong** [comment]
Wondering if it'd be better to wait till .Net 5 comes out. Though [omnisharp-roslyn](https://github.com/OmniSharp/omnisharp-roslyn) has linux-arm64 support, so it should be doable. Unfortunately linux dev isn't really my forte so I haven't been able to get it to build.

**Comment by SLS8** [comment]
> Wondering if it'd be better to wait till .Net 5 comes out.
Don't wait, change the software. Use Visual Studio Community Edition (or Remote Desktop to a PC or Azure with it) to develop C# code then only run it on Raspberry.  As the message says, the c# extension is fully "incompati**a**ble"

**Comment by scottyearsley** [comment]
> > Wondering if it'd be better to wait till .Net 5 comes out.
> 
> Don't wait, change the software. Use Visual Studio Community Edition (or Remote Desktop to a PC or Azure with it) to develop C# code then only run it on Raspberry. As the message says, the c# extension is fully "incompati**a**ble"

For me and obviously others, this would be great to get working - I have a number of PCs with VS, but I like to take my Chromebook with me and like to tinker with code whilst having a coffee out (not so much at the mo!) or just sat in front of the TV. With Apple moving to ARM and the boom of Chromebooks being used as a primary device, I do not see this as being a need for fringe users at all.

**Comment by SLS8** [comment]
I understand, I'd like to use VS Code and C# on Raspberry 4, but without intellisense it's not usable. So don't wait for .net 5, I doubt  the c# extension is going to be "compatiable" anytime soon.

**Comment by baralong** [comment]
> > Wondering if it'd be better to wait till .Net 5 comes out.
> 
> Don't wait, change the software. Use Visual Studio Community Edition (or Remote Desktop to a PC or Azure with it) to develop C# code then only run it on Raspberry. As the message says, the c# extension is fully "incompatiable"

That doesn't mean it'll never be compatible there's a prerelease omnisharp-linux-arm64-1.34.15.zip listed here "https://roslynomnisharp.blob.core.windows.net/releases?restype=container&comp=list" though its length is 0, but it looks like it's on the road map

**Comment by mmundy3832** [comment]
Put me down as another person requesting this. It seems like a simple decision to get going on this feature, especially now that the RPi 4 is out, as well as so many other SBCs going to 64 bit.

I actually tried installing the 32-bit version of VSCode hoping that would get around the issue, but alas. I still got the incompatibility message. Unfortunately, the linked "work around" page was not terribly helpful in helping me work around this issue.

**Comment by modul8com** [comment]
Please create a Linux ARM64 (and Linux ARM32) version of OmniSharp, so that we can use the Pi 400 for .NET 5 educational purposes !
I mean VS Code works now on the Pi 4(00), and .NET 5 also runs on Linux ARM32 and Linux ARM64... we just can't use the C# plugin in VS Code since OmniSharp doesn't support the Linux ARM architecture at all.
Ask the guys from the VS Code team how to do it, they succeeded finally.
I'd be willing to help out, but I have no clue how to do it :(.

**Comment by archanox** [comment]
Is there anything that the community do to help this progress? Are there any blockers at the moment?
Perhaps crowd funding to get this re-prioritised?

**Comment by lokinfey** [comment]
I wait got  arrch64 version . When I can get that ......

**Comment by Cyber1000** [comment]
I've found a way to get debugging working on my rpi4-64bit (aarch64) at least with dotnet-5.0, should work on rpi with 32 bit too (and maybe other linux ARM versions - not sure if I caught all possible architectures) - see PR #4515 - testers welcome

**Comment by modul8com** [comment]
Whaaaaat so simple ? Cool that you fixed it... I hope it gets quickly merged. I tweeted Scott Hanselman about it, so it might actually happen 🤤

**Comment by Cyber1000** [comment]
Well I was surprised too :sweat_smile: , everything seemed to be in it's place, but the checks were working against it.

**Comment by modul8com** [comment]
How can I test it, I've cloned the repo on my PC, did the npm i and npm run compile, and now I have an out folder... how do I test it on the Pi, just copy the out folder over to somewhere on the pi ?  But where ?  I searched on the PC in the extensions folder, but there is no omnisharp there.

**Comment by Cyber1000** [comment]
I did this on a commandline:
```
npm install –g vsce
vsce package
```

- First command: maybe elevation is needed, cause of the global install of vsce
- run second command from within the checked out omnisharp code, it creates a vsix-package, at the end you will get the path where it is stored
- you can copy over it now and install it on your arm device (maybe you need to uninstall the offical omnisharp before installing it as vsix-package)

After running "vsce package":
![image](https://user-images.githubusercontent.com/1142765/116005516-95044900-a607-11eb-86df-43e9fb8fd0a4.png)

Installing from vsix:
![image](https://user-images.githubusercontent.com/1142765/116005564-d0067c80-a607-11eb-88af-5b365df071e6.png)

**Comment by Cyber1000** [comment]
And for the sake of completeness:
- if someone uses docker instead you can include a selfbuild vsix-File with copying in your dev-folder and adding these lines in devcontainer.json :
```
	"extensions": [
		"/workspace/csharp-1.23.11.vsix"
	],
```

**Comment by Aemilivs** [comment]
@Cyber1000 applied small changes to your pr, please review:
https://github.com/Cyber1000/omnisharp-vscode/pull/1

**Comment by Aemilivs** [comment]
@Cyber1000 
Can't get it to work on RPi 4 (installed the built vsix package locally and opened a dotnet 5 solution via ssh on a laptop)

**Comment by Cyber1000** [comment]
@Aemilivs With "installed locally" you don't mean your rpi4 but your laptop, if I read that right? And the code you want to debug resides on the rpi4.
I'm using it within docker running on my rpi4, and the plugin is installed directly in docker on my rpi4 (sharing the same kernel, so same arch)

- you ssh from your laptop to rpi4
- you may look in your extensions now if the offical omnisharp is already installed on the remote end. If yes remove it. (In my case only a docker-addon is installed on the rpi4 itself)
![image](https://user-images.githubusercontent.com/1142765/116159619-dfadc000-a6f0-11eb-801d-c935fbf8d1e4.png)
- you may need to restart vscode and/or reconnect the ssh 
- when connected/reconnected: F1 and "Install from vsix"
- you now search your filesystem on the rpi4 (you should copy over the vsix before, wherever you created it)
- you may need to restart vscode and/or reconnect the ssh before you can use it

In my post yesterday I assumed the easiest case - building the vsix locally on an arm device and using it locally on the same arm device. SSHing into another machine is like sitting in front of the other machine, the plugin needs to be there.

**Comment by Aemilivs** [comment]
@Cyber1000 `With "installed locally" you don't mean your rpi4 but your laptop`
No, I mean rpi4 :)

I don't know about building the plugin on rpi4, but I will try to the docker setup that you described, thanks 👍

**Comment by Cyber1000** [comment]
> Can't get it to work on RPi 4 (installed the built vsix package locally and opened a dotnet 5 solution via ssh on a laptop)

- Thanks for clarification, that just sounded as if your dotnet-solution is not directly on your rpi4 and you try to debug somewhere else on a non arm64 device.
- I don't think that this plugin is depending on a specific arch, since it's typescript, I created it on my x86 laptop and just copied it to a docker-container on my rpi4 with aarch64.
- docker shouldn't make a big difference, since it shares the same host kernel, maybe it depends on architecture, I'm using 64bit raspberry OS, which may be still in beta. Though my PR should add 32bit arm too ...

**Comment by davebrunger** [comment]
@Cyber1000 I built and installed the plugin using your instructions on my Raspberry Pi 4 8GB running Ubuntu 64 bit. Debugging C# works a treat, but is a little slow, but I don't have a great cooling solution. Fantastic job, well done. I hope you get the Pull Request approved.

Now, how do I get the C# code completion to work?...

**Comment by Cyber1000** [comment]
I think code completion worked, not quite sure. Had some other work in between, can have a look on this at the weekend.

**Comment by baralong** [comment]
> I did this on a commandline:
> 
> ```
> npm install –g vsce
> vsce package
> ```

I've been following this for working on my arm powered chromebook. My processor reports as 'aarch64' 

I added that to 'src/main.ts' but it reports that debugging isn't available, and it reports the Omnisharp server isn't running. Any tips would be appreciated.

**Comment by Cyber1000** [comment]
What did you add to 'src/main.ts'? All the things from my PR (it includes aarch64 too, I tested it with this architecture)? https://github.com/OmniSharp/omnisharp-vscode/pull/4515/files
Also adjusted activate.ts and package.json?

**Comment by baralong** [comment]
> What did you add to 'src/main.ts'? All the things from my PR (it includes aarch64 too, I tested it with this architecture)? https://github.com/OmniSharp/omnisharp-vscode/pull/4515/files
> Also adjusted activate.ts and package.json?

🤦 Sorry, I pulled your repo and then built the wrong branch. Just tried again with the 'armv7l_aarch64' branch built and and it installed correctly. I'll do some further testing.

**Comment by FTomaYoamb** [comment]
I've installed the vsix package from the "armv7l_aarch64" branch (uninstalled existing one, reload, installed vsix) on my arm64 chromebook but I get this error : 
_`[ERROR] Error: spawn /home/florin/.vscode/extensions/ms-dotnettools.csharp-1.23.11/.omnisharp/1.37.8/run ENOENT`_
that ".omnisharp" folder doesn't seem to be there.

**Comment by baralong** [comment]
> I've installed the vsix package from the "armv7l_aarch64" branch (uninstalled existing one, reload, installed vsix) on my arm64 chromebook but I get this error :
> _`[ERROR] Error: spawn /home/florin/.vscode/extensions/ms-dotnettools.csharp-1.23.11/.omnisharp/1.37.8/run ENOENT`_
> that ".omnisharp" folder doesn't seem to be there.

**Comment by baumhoto** [comment]
Got code completion to work on my Raspi Pi 4  by manually installing the omnisharp server (mono version as there is no arm build available). I’ve followed the instructions from Cyber1000 above to build the vscode-omnisharp extension (I’ve updated it to 1.23.12 though) and installed the extension. Code completion was not working and i got the error about the missing .omnisharp folder like others above.

To get code-completion to work i did the following:

1. Follow instructions to install latest mono stable: [https://www.mono-project.com/download/stable/#download-lin-raspbian](https://www.mono-project.com/download/stable/#download-lin-raspbian)
2. Make sure mono is working `mono —version`
2. Goto the C# extension folder `cd ~/.vscode/extensions/ms-dotnettools.csharp-1.23.12`  (Might be a different version-number at the end in your case)
3. Create the .omnisharp subfolder `mkdir .omnisharp/1.37.10/omnisharp`
4. Change into subfolder `cd .omnisharp/1.37.10/omnisharp`
5. Download omnisharp mono `wget https://github.com/OmniSharp/omnisharp-roslyn/releases/download/v1.37.10/omnisharp-mono.tar.gz`
6. Extract it: `tar xzf omnisharp-mono.tar.gz`
7. In VsCode goto Prefences->Settings and search for `use global mono`
8. Change it from `auto` to `always`
9 . Code completion should now work

**Comment by sanme98** [comment]
> I've installed the vsix package from the "armv7l_aarch64" branch (uninstalled existing one, reload, installed vsix) on my arm64 chromebook but I get this error :
> _`[ERROR] Error: spawn /home/florin/.vscode/extensions/ms-dotnettools.csharp-1.23.11/.omnisharp/1.37.8/run ENOENT`_
> that ".omnisharp" folder doesn't seem to be there.

Any help?

**Comment by SuperJMN** [comment]
I can't believe this is still not there.

**Comment by skreimeyer** [comment]
This is a feature I would also like to see. I was disappointed to find that omnisharp is not available on raspberry pi. Ionide adds a lot of value to VSCode for F#.

**Comment by Dean2678** [comment]
Hi @Cyber1000, @baumhoto 

Really appreciate your efforts on this! 

I've followed both of your instructions to the letter on a fresh Rasbian 10 build, but when I try to debug a simple console app I have the following message:

"The C# extension for Visual Studio Code (powered by Omnisharp) is incompatible on Linux aarch64

Source: C# (Extension)" 

Any ideas? Is there a working build of the VSIX file I could download and try please, in case something went wrong while I was packaging it up?

Any help greatly appreciated!

**Comment by Cyber1000** [comment]
@Dean2678 Glad you found a solution, sorry didn't read your comment in time. 

@sandrokl I've only heard this https://www.infoworld.com/article/3608611/whats-new-in-microsoft-net-6.html
> Support for MacOS and Windows Arm64 is almost complete. .NET 6 RC2 enables Arm64 + x64 coexistence by installing Arm64 and x64 builds to different locations.

But there are hopefully others who know more ...

---

For everyone who wants to test, i've updated my fork and with an github-action (which is also used here) I built a new vsix: https://github.com/Cyber1000/omnisharp-vscode/releases/tag/v1.23.17-arm64.alpha

For now I have one positive reply (from myself) and one negative (see in PR #4515), would be interested if it works (or doesn't work) for others.
My setup: win10 connected to docker via docker-context/ssh (docker running on rpi4 64bit - extension installed only in container)

**Comment by JoeRobich** [comment]
A prerelease build of the C# extension with experimental language server support for Linux Arm64 is available at https://github.com/OmniSharp/omnisharp-vscode/releases/tag/v1.23.18-beta1. You will need to set your "omnisharp.path" to "latest" as there is not a published release of O# for this platform. There is also no Razor language support in this build.

**Comment by mmundy3832** [comment]
I've been away from the thread for a while, but I certainly appreciate the effort to getting this new support into the project. I look forward to testing it in the new year when I can get back to the project that got me into this thread.

**Comment by WoojO** [comment]
What about armv7l and raspberry pi 4? Is there plan to update plugin?

**Comment by jtheisen** [comment]
I struggling to figure out what raspi omnisharp works on.

I have a Raspi 4 and `cat /proc/cpuinfo` tells me

```
Model : Raspberry Pi 4 Model B Rev 1.4
```

but also

```
model name : ARMv7 Processor rev 3 (v7l)
```

and so omnisharp doesn't work.

According to online sources, Raspi 4s should have ARMv8. What raspi can I use?

**Comment by JoeRobich** [comment]
@jtheisen I believe the default Raspian distro is a 32-bit OS. You could install a 64-bit Debian distro instead.

**Comment by sanme98** [comment]
@jtheisen, what JoeRobich said is correct, default Raspian is 32-bit OS. You can't get the image from Raspberry Pi Imager directly because it is still in beta > https://forums.raspberrypi.com/viewtopic.php?t=275370

**Comment by m-brooks** [comment]
You can get the latest 64 bit Raspbian build from here: https://downloads.raspberrypi.org/raspios_arm64/images/

**Comment by andrew-benson** [comment]
> A prerelease build of the C# extension with experimental language server support for Linux Arm64 is available at https://github.com/OmniSharp/omnisharp-vscode/releases/tag/v1.23.18-beta1. You will need to set your "omnisharp.path" to "latest" as there is not a published release of O# for this platform. There is also no Razor language support in this build.

Was hoping this woud work for my Chromebook (aarch64) Below is what I'm greeted with in the output window. Tried a number of solutions on the internet to rectify this GLIBC not found issue but no dice. I'm guessing this was made with Raspberry Pi in mind

`Starting OmniSharp server at 12/14/2021, 6:20:39 PM
    Target: /home/abenson/Projects/HelloWorld

OmniSharp server started.
    Path: /home/abenson/.vscode/extensions/ms-dotnettools.csharp-1.23.18/.omnisharp/1.37.18-beta.115/run
    PID: 13010

/home/abenson/.vscode/extensions/ms-dotnettools.csharp-1.23.18/.omnisharp/1.37.18-beta.115/bin/mono: /lib/aarch64-linux-gnu/libm.so.6: version `GLIBC_2.29' not found (required by /home/abenson/.vscode/extensions/ms-dotnettools.csharp-1.23.18/.omnisharp/1.37.18-beta.115/bin/mono)`

**Comment by andrew-benson** [comment]
> > A prerelease build of the C# extension with experimental language server support for Linux Arm64 is available at https://github.com/OmniSharp/omnisharp-vscode/releases/tag/v1.23.18-beta1. You will need to set your "omnisharp.path" to "latest" as there is not a published release of O# for this platform. There is also no Razor language support in this build.
> 
> Was hoping this woud work for my Chromebook (aarch64) Below is what I'm greeted with in the output window. Tried a number of solutions on the internet to rectify this GLIBC not found issue but no dice. I'm guessing this was made with Raspberry Pi in mind
> 
> `Starting OmniSharp server at 12/14/2021, 6:20:39 PM Target: /home/abenson/Projects/HelloWorld
> 
> OmniSharp server started. Path: /home/abenson/.vscode/extensions/ms-dotnettools.csharp-1.23.18/.omnisharp/1.37.18-beta.115/run PID: 13010
> 
> /home/abenson/.vscode/extensions/ms-dotnettools.csharp-1.23.18/.omnisharp/1.37.18-beta.115/bin/mono: /lib/aarch64-linux-gnu/libm.so.6: version `GLIBC_2.29' not found (required by /home/abenson/.vscode/extensions/ms-dotnettools.csharp-1.23.18/.omnisharp/1.37.18-beta.115/bin/mono)`

Okay so to answer my own problem (and for anyone out there with a chromebook) I did some research and it appears it's because the libc/glibc version on my chromebook is below 2.29. I used `dpkg -l libc6` to find that it was 2.28-10 (so close). Apparently you can upgrade the libc version but it's been compared to a brain transplant so it's advised not to do so.

**Comment by m-brooks** [comment]
It is now just over three and a half years since I posted the OP. I have finally found time to have another go at getting this all set up on a Raspberry Pi running Raspbian and I have to say I am delighted with the results. For some reason actually getting .Net 6 properly installed took some work (the install script didn't appear to succeed at setting environment variables appropriately - I had to manually update .profile to set PATH and DOTNET_ROOT), but once I nailed that I found that everything works. I have .Net 6, dotnet-script, VSCode, the C# extension, source-level debugging and Intellisense - even when using .csx scripts (which was what I was really after).
![20220818_194107](https://user-images.githubusercontent.com/35042732/185479570-4a7d327e-2a54-4b0d-a606-6e73c81dcfa3.jpg)
Amazing work by everyone involved. We now have what I consider to be an awesome low cost educational package using a proper language. :-)

**Comment by changetocoding** [comment]
Can anyone help me: I still get that warning plus 'Couldn't find a debug adapter descriptor for debug type coreclr'

Here's the steps I've taken (on a raspberry pi)

1. Installed Visual Studio
2. Installed .net6 - https://dotnet.microsoft.com/en-us/download/dotnet/thank-you/sdk-6.0.400-linux-arm32-binaries
3. Created a new console app c# project: https://dotnet.microsoft.com/en-us/learn/dotnet/hello-world-tutorial/edit
4. Compiled and run it in the terminal to ensure working: This works fine and displays hello world on the console.
```
dotnet run
```
5. Downloaded Omnisharp v1.23.18 beta - https://github.com/OmniSharp/omnisharp-vscode/releases/tag/v1.23.18-beta1
6. Installed extension from VSIX in VsCode - https://code.visualstudio.com/docs/editor/extension-marketplace#_install-from-a-vsix
7. Opened folder in vs code
8. Added a launch.json

I initially tried a few other things. Fresh install maybe needed?
Do I need a 64-bit distro? Not sure what my distro is but I guess it's a 32 bit.
Do I need to update my raspberry pi distro?

Thanks,

**Comment by changetocoding** [comment]
@JoeRobich Thanks :)

**Comment by dibarbet** [comment]
Since this issue was filed O#/C#/Dev Kit all now support linux ARM.  Closing

**Comment by JoseMPersichini** [comment]
I'm trying to debug a lambda function using AWS Toolkit and I'm getting the error

`The vsdbg debugger does not currently support the arm64 architecture. Function will run locally without debug. Source: AWS Toolkit.
`
It works, but it doesn't stop at breakpoints. I have an Apple M2 chip. I'm reopening the issue because it's pointed on the [AWS docs](https://docs.aws.amazon.com/toolkit-for-vscode/latest/userguide/appbuilder-overview-overview.html)

**Comment by JoeRobich** [comment]
@JoseMPersichini Could you open a new issue and provide the requested logs?
